### PR TITLE
FFI extra safety bounds 

### PR DESCRIPTION
**Pull request by SWvheerden**
Check all FFI calls where possible to ensure that the calls into FFI didnt fail before we return to safe rust

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ pub enum RandomXFlag {
 }
 
 impl RandomXFlag {
-    pub fn value(self) -> c_uint {
+    pub fn value(self) -> u32 {
```
Comment (StriderDM) [inline src/lib.rs]: This value gets passed to C, while equivalent at present is u32 it is probably best left as a c_uint

File: src/lib.rs [hunk 2]
```diff
@@ -108,6 +108,7 @@ impl RandomXCache {
             let key_ptr = key.as_bytes().as_ptr() as *mut c_void;
```
Comment (StriderDM) [inline src/lib.rs]: Needs a check to ensure length of key is larger than zero.

File: src/lib.rs [hunk 3]
```diff
@@ -154,23 +155,30 @@ impl RandomXDataset {
...
             }
             Ok(result)
         }
     }
 
-    pub fn count(&self) -> c_ulong {
-        unsafe { randomx_dataset_item_count() }
+    pub fn count(&self) -> Result<u64, CreationError> {
+        match unsafe { randomx_dataset_item_count() } {
+            0 => Err(CreationError),
```
Comment (StriderDM) [inline src/lib.rs]: I don't think the return of zero is necessarily an error here.

File: src/lib.rs (hunk 1 above)
Comment (SWvheerden) [reply src/lib.rs]: I see every where it gets passed to randomX, it gets cast as a c_uint first. 

From the rust docs: 
```
Equivalent to C's unsigned int type.

This type will almost always be u32, but may differ on some esoteric systems. The C standard technically only requires that this type be an unsigned integer with the same size as an int; some systems define it as a u16, for example.
```

So I think if we keep it a u32 on our side, and then casting it to c_unit before we use it, it should be safe everywhere.

File: src/lib.rs (hunk 2 above)
Comment (SWvheerden) [reply src/lib.rs]: good catch

File: src/lib.rs (hunk 3 above)
Comment (SWvheerden) [reply src/lib.rs]: I might not be an error. But its very likely that the dataset item count will never be 0. 
0 is the default u64 value. Thats why I am taking a guess that if this is 0, it failed and the u64 has its default value. 

Do you think this is perhaps too dangerous?

Comment (StriderDM) [reply src/lib.rs]: I am inclined to agree with you that this in all probability should not be zero.

File: src/lib.rs [hunk 4]
```diff
@@ -153,24 +157,39 @@ impl RandomXDataset {
...
-    pub fn get_data(&self) -> Vec<u8> {
+    pub fn get_data(&self) -> Result<Vec<u8>, CreationError> {
         let memory = unsafe { randomx_get_dataset_memory(self.dataset) };
+        if memory.is_null() {
+            return Err(CreationError);
+        }
         let mut result = Vec::new();
         for i in self.dataset_start..self.dataset_count {
             result.push(unsafe { memory.offset(i as isize) } as u8);
         }
```
Comment (CjS77) [inline src/lib.rs]: I think safe code should check that the following contracts are honoured:

`dataset_start` and `dataset_count` don't ever let us access out of bounds memory.

We could also probably >> performance by moving this into one unsafe block; allocating the correct amount of memory at once (`Vec::new()` will lead to many re-allocations and unneeded copies) - punted to #9

File: src/lib.rs [hunk 5]
```diff
@@ -210,31 +229,40 @@ impl RandomXVM {
...
 
     pub fn reinit_dataset(&self, dataset: &RandomXDataset) {
+        //no way to check if this fails, c code does not return anything
         unsafe {
             randomx_vm_set_dataset(self.vm, dataset.dataset);
         }
     }
 
-    pub fn calculate_hash(&self, input: &str) -> Vec<u8> {
+    pub fn calculate_hash(&self, input: &str) -> Result<Vec<u8>, CreationError> {
```
Comment (CjS77) [inline src/lib.rs]: `CreationError` does not sound like the correct Error type here. 

At this point, the VM has been created and we're checking hashes / mining with it

File: src/lib.rs [hunk 6]
```diff
@@ -289,7 +317,9 @@ mod tests {
         flags.push(RandomXFlag::FlagDefault);
         let cache = RandomXCache::new(flags.clone(), key).unwrap();
         let dataset = RandomXDataset::new(flags.clone(), &cache, 0).unwrap();
-        let memory = dataset.get_data();
+        let memory_result = dataset.get_data();
+        assert_eq!(memory_result.is_ok(), true);
+        let memory = memory_result.unwrap();
```
Comment (CjS77) [inline src/lib.rs]: `let memory = dataset.get_data().expect("...");` does this in one line.

File: src/lib.rs (hunk 4 above)
Comment (SWvheerden) [reply src/lib.rs]: dataset_start and dataset_count is checked when self is/was created by safe code. At this point it should be safe. 
If memory comes back as not null it should also be safe

File: src/lib.rs (hunk 6 above)
Comment (SWvheerden) [reply src/lib.rs]: updated

File: src/lib.rs (hunk 5 above)
Comment (SWvheerden) [reply src/lib.rs]: I will recreate all the error class

File: src/lib.rs (hunk 4 above)
Comment (CjS77) [reply src/lib.rs]: Assigning the values in safe code doesn't make this safe; you have to provide the guarantees yourself as per the Rustinomicon. That said, I looked again, and the bounds are checked in line #159, so I think we're good.

**Pull request merged**